### ISSUE TITLE
unexpected `pnpm deploy` behavior with localized gitignore

### ISSUE THREAD

**Issue by josefaidt** [pnpm/pnpm#7286]
### Contribution

- [X] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

### Describe the user story

As a developer, I want to maintain localized gitignore files for different frameworks in my monorepo. After installing dependencies and building the target workspace package, I deploy with `pnpm deploy --filter ./apps/my-astro-site`, however the `dist/` directory is not copied. However if I ignore the dist directory from my project root's gitignore, the `dist/` directory is copied.

By co-opting the localized gitignore for `pnpm deploy`'s ignore it produces unexpected behavior by ignoring built assets of workspace packages.

### Describe the solution you'd like

- remove behavior reading localized `gitignore` for `pnpm deploy`
- optionally introduce a CLI option for `pnpm deploy` to specify ignored directories, files
- optionally introduce a config option in `pnpm-workspace.yaml` to specify deploy behavior for workspace packages (by filename, glob, etc.)

### Describe the drawbacks of your solution

- removing behavior reading localized `gitignore` for `pnpm deploy` seems like a breaking change for those relying on this behavior in their pipelines
- introducing a CLI option needs proper documentation for discoverability
- introducing a config option in the existing workspace file may overload this file's intent, could break if folks are switching between projects with an older version of pnpm specified in `packageManager`

### Describe alternatives you've considered

if the behavior is desired, documenting it in the official pnpm docs

**Comment by martinc1991** [comment]
Spent 12 hours today to build a single docker image, please document it at least.

**Comment by rohanrajpal** [comment]
> ### Contribution
> * [x]  I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))
> 
> ### Describe the user story
> As a developer, I want to maintain localized gitignore files for different frameworks in my monorepo. After installing dependencies and building the target workspace package, I deploy with `pnpm deploy --filter ./apps/my-astro-site`, however the `dist/` directory is not copied. However if I ignore the dist directory from my project root's gitignore, the `dist/` directory is copied.
> 
> By co-opting the localized gitignore for `pnpm deploy`'s ignore it produces unexpected behavior by ignoring built assets of workspace packages.
> 
> ### Describe the solution you'd like
> * remove behavior reading localized `gitignore` for `pnpm deploy`
> * optionally introduce a CLI option for `pnpm deploy` to specify ignored directories, files
> * optionally introduce a config option in `pnpm-workspace.yaml` to specify deploy behavior for workspace packages (by filename, glob, etc.)
> 
> ### Describe the drawbacks of your solution
> * removing behavior reading localized `gitignore` for `pnpm deploy` seems like a breaking change for those relying on this behavior in their pipelines
> * introducing a CLI option needs proper documentation for discoverability
> * introducing a config option in the existing workspace file may overload this file's intent, could break if folks are switching between projects with an older version of pnpm specified in `packageManager`
> 
> ### Describe alternatives you've considered
> if the behavior is desired, documenting it in the official pnpm docs

A bit confused, so you're saying if you ignore the `dist/` in root gitignore it copies only then? How exactly is pnpm deploy and gitignore relates?

I'm kind of also stuck where it seems my prisma generated files are not being copies and localized `gitignore` has mentioned `src/generated/client` but its not in the root gitignore, that means it wont be copies unless i mention it in the root `gitignore`? why is gitignore being read anyway?

**Comment by antiBaconMachine** [comment]
**tl;dr; Workaround is to put an empty .npmignore in your application directory**

I ran into this today, it cost a good few hours to figure out what is going on. The original report above is pretty solid but in case it helps clarrify things, I've made a test repo which shows how files, .npmignore and .gitignore all affect the packlist which deploy uses https://github.com/antiBaconMachine/pnpm-deploy-packlist-examples/tree/main.

@rohanrajpal asks a good question:
> How exactly is pnpm deploy and gitignore related?

Which I echo. I _think_ they are related becuase pnpm uses or at least mimics https://www.npmjs.com/package/npm-packlist. See the readme for a complete description but essentially this package will build the packlist in this order:

1. If there is a files property in package.json then explicitly include only these files
2. If there is an .npmignore file in the application dir then copy everything except files listed in there
3. If there is a .gitignore file in the application dir then copy everything except files listed in there

This is demonstrated in the test repo https://github.com/antiBaconMachine/pnpm-deploy-packlist-examples/tree/main

So a workaround for this issue is to simply put an empty `.npmignore` file in your application directory and then this will be used in preference to the .gitignore file.

Agree, this should _absolutely_ be in the docs as currently they just say

>By default, all the files of the project are copied during deployment. The project's package.json may contain a "files" field to list the files and directories that should be copied.

Which is a very incomplete description of how this works. I have opened a docs PR over here https://github.com/pnpm/pnpm.io/pull/559 to clarrify the situation.

**Comment by zkochan** [comment]
I am sorry, I don't think files in gitignore should be excluded from deploy. If that's how it currently works, we should fix it

**Comment by KSXGitHub** [comment]
This behavior is inherited from `npm-packlist` which is also used in `npm pack`. The behavior is as follow: If `package.json` doesn't define a file list and the package root directory doesn't have a `.npmignore` file, it will treat `.gitignore` file as the ignore file.

So the workaround is to create a local `.npmignore` file.

I don't know whether pnpm should deviate from standard npm behavior.
